### PR TITLE
Auto device fallback on provider failure and tokenizer_options passthrough in Text2TextGenerationPipeline

### LINKED ISSUE #425: [Feature request] Return offset mapping using tokenizer
**Return offset mapping using tokenizer**

Hi guys, awesome work with this project, it is helping us a lot! Is it possible to add the parameter "return_offset_mapping" to the Tokenizer class, just like in the Transformers python library?

**Reason for request**

The returned offsets are crucial for our project in Named Entity Recognition, as we need to associate the predicted tags with the original sentence (prior to tokenization). Currently we can only associate the predicted tags with the list of subtokens, as they are a 1-to-1 association, and we know that the list of subtokens ignore some elements, such as whitespaces, which are present in the original sentence.

**Additional context**

Reference for "return_offset_mapping" on Transformers python library: https://huggingface.co/docs/transformers/v4.35.2/en/main_classes/tokenizer#transformers.PreTrainedTokenizerFast.__call__.return_offsets_mapping

### LINKED ISSUE #633: Is 'aggregation_strategy' parameter available for token classification pipeline?
### Question

Hi. I have question.

From HuggingFace Transformers documentation, they have **'aggregation_strategy'** parameter in token classification pipeline. [Link](https://huggingface.co/docs/transformers/en/main_classes/pipelines#transformers.TokenClassificationPipeline.aggregation_strategy)
Need to know in this library provide this parameter?

Thanks.

### LINKED ISSUE #1096: Allow configuring the tokenizer in `Text2TextGenerationPipeline`
### Feature request

The current behavior has hardcoded options for the tokenizer https://github.com/huggingface/transformers.js/blob/main/src/pipelines.js#L761-L775

### Motivation

It would be nice to stick with the generic `pipeline` API and be able to pass through custom tokenizer options that are overwriting `tokenizer.json`

### Your contribution

If the idea sounds good I can contribute the patch

### LINKED ISSUE #1642: Auto device on Linux x64 fails hard when CUDA shared library is unavailable instead of falling back
### System Info

# Minimal repro

Environment
* OS: Linux x64
* CPU: AMD Ryzen (no NVIDIA CUDA runtime installed)
* Node: LTS 24.14.0
* @huggingface/transformers: 4.0.1 (or exact resolved version)

### Environment/Platform

- [ ] Website/web-app
- [ ] Browser extension
- [x] Server-side (e.g., Node.js, Deno, Bun)
- [ ] Desktop app (e.g., Electron)
- [ ] Other (e.g., VSCode extension)

### Description

# Summary

When running in Node on Linux x64 with device set to `'auto'`, Transformers.js currently prioritizes `'cuda'` and fails with a hard error if CUDA shared libraries are not available. This prevents fallback to other providers (like `'webgpu'` or `'cpu'`).

# Observed error

```
Error: OrtSessionOptionsAppendExecutionProvider_Cuda: Failed to load shared library
      at new OnnxruntimeSessionHandler (node_modules/onnxruntime-node/lib/backend)
      at Immediate.<anonymous> (node_modules/onnxruntime-node/lib/backend.ts:147:)
      at process.processImmediate (node:internal/timers:504:21)
```

On Linux x64, CUDA is added to supported devices first in `onnx.js`:

```js
    switch (process.platform) {
        case 'win32': // Windows x64 and Windows arm64
            supportedDevices.push('dml');
            break;
        case 'linux': // Linux x64 and Linux arm64
            if (process.arch === 'x64') {
                supportedDevices.push('cuda');
            }
            break;
        case 'darwin': // MacOS x64 and MacOS arm64
            supportedDevices.push('coreml');
            break;
    }

    supportedDevices.push('webgpu');
    supportedDevices.push('cpu');
    defaultDevices = ['cpu'];
```

# Proposed fix

In Node, when `device` is `'auto'` and session creation fails specifically due to CUDA provider shared-library load error, retry once with executionProviders filtered to remove cuda.

Even better: `'auto'` should retry the list from the best performance preferred first to safe fallback option ('cpu').

`'auto'` returns that list which is passed directly to session options in session.js.

### Reproduction

# Run this snippet on an AMD APU

```js
import { pipeline } from '@huggingface/transformers'

await pipeline('text-generation', 'onnx-community/LFM2-1.2B-ONNX', {
  device: 'auto',
  dtype: 'q4',
})
```

# Expected behavior

With `device` set to `'auto'`, unavailable CUDA provider should be skipped and session creation should continue with remaining providers (for example webgpu or cpu).

# Actual behavior

Session creation fails immediately on CUDA provider shared-library error.

### LINKED ISSUE #1643: get_available_devices()
### Feature request

Add a function to get all the possible valid values of `device` in a given environment (e.g. node, electron, browser).

### Motivation

We have `ModelRegistery.get_available_dtypes()` which is a huge help in user onboarding:

- Instead of guessing the best dtype for a given machine, we can let the users choose.
- We can try different values and see which one works (fallback mechanism).

But when it comes to `device`, we're left with guessing.

- `auto` has its own problems (#1624)
- Without a list, the code for trying fallbacks relies on hard coded strings

The value is available as `suportedDevices` in [onnx.js](https://github.com/huggingface/transformers.js/blob/main/packages/transformers/src/backends/onnx.js) but not exposed externally.

### Your contribution

I could implement a PR if needed.

### LINKED ISSUE #1666: Gemma 4 generation passes num_logits_to_keep=0 in decoder_forward, causing full-prompt logits memory blowup
### Bug

Gemma 4 generation can pass `num_logits_to_keep=0` into the ONNX decoder session, which causes full-prompt logits to be computed during generation. For long prompts this wastes a large amount of memory and can contribute to OOMs.

### Evidence from current `main`

In `packages/transformers/src/models/modeling_utils.js`, `decoder_prepare_inputs_for_generation()` correctly sets `num_logits_to_keep=1` for generation:

```js
// During generation, only the last token's logits are needed. Setting num_logits_to_keep=1
// avoids computing logits for the entire sequence, significantly reducing memory usage.
const session = self.sessions['decoder_model_merged'] ?? self.sessions['model'];
if (session?.inputNames.includes('num_logits_to_keep') && !model_inputs.num_logits_to_keep) {
    model_inputs.num_logits_to_keep = new Tensor('int64', [1n], []);
}
```

But `decoder_forward()` has the same comment and then inserts `0n` when the input is missing:

```js
if (session.inputNames.includes('num_logits_to_keep') && !new_model_inputs.num_logits_to_keep) {
    // ... During generation, the default is 1 ...
    new_model_inputs.num_logits_to_keep = new Tensor('int64', [0n], []);
}
```

For Gemma 4/Gemma3n conditional generation, `Gemma3nForConditionalGeneration.forward()` calls `decoder_forward(this, ..., true)` directly. In practice the ONNX feed reaches `decoder_model_merged` with `num_logits_to_keep=0`.

### Reproduction context

We hit this in [Wandler](https://wandler.ai), an OpenAI-compatible local inference server powered by Transformers.js, while running `onnx-community/gemma-4-E4B-it-ONNX` with `@huggingface/transformers@4.2.0`.

Wandler tracking issue for removing our temporary workaround once this is fixed upstream:

- https://github.com/runpod-labs/wandler/issues/40

The same code path is relevant to the promoted Gemma 4 browser assistant pattern using:

```js
pipeline("text-generation", "onnx-community/gemma-4-E2B-it-ONNX", {
  dtype: "q4f16",
  device: "webgpu",
})
```

### Measured impact

For a ~20k token Gemma 4 prompt:

- vocab size: 262144
- full-prompt logits estimate: ~20 GB
- after forcing `num_logits_to_keep=1` at the ONNX feed, that logits waste disappears

We also separately found a full-attention prefill allocation issue for long prompts, but this issue is only about the logits side: generation should not request full prompt logits.

### Expected behavior

During generation, Gemma 4 decoder sessions with `num_logits_to_keep` should receive `1`, not `0`, unless the user explicitly requested full-sequence logits.

### Possible fix

`decoder_forward()` should not default `num_logits_to_keep` to `0n` for generation paths. It likely needs to preserve an explicit user value but use `1n` when called from generation. At minimum, the current comment and behavior conflict.

### Version checked

- `@huggingface/transformers`: 4.2.0
- upstream `main` inspected at commit `f7487c737aa8cafbc106c9adf69dc9578c8f3fe0`

### PR DESCRIPTION
**Pull request by Suh0161**
Closes #425, closes #633, closes #1643, closes #1666, closes #1642, closes #1096.

Supersedes #1668 and #1669 (same changes, single PR).

### Summary

- **#425 / #633** — `return_offsets_mapping` on the tokenizer; token-classification outputs can include `start` / `end` character spans; GPT-2 ByteLevel `offset_mapping` includes the leading space where appropriate.
- **#1643 / #1666** — Adds `get_available_devices()`; fixes default `num_logits_to_keep` in `decoder_forward` so it matches generation behavior and avoids computing full-prompt logits (memory / OOM risk on large vocab + long prompts).
- **#1642 / #1096** — When `device: 'auto'`, tries execution providers sequentially instead of crashing if a higher-priority provider (e.g. CUDA) fails to load; `tokenizer_options` in `Text2TextGenerationPipeline` is forwarded to the tokenizer and stripped before `generate()`.

### Fix: Auto device falls back gracefully when a provider fails to load (#1642)

When `device: 'auto'` is used on a machine where a high-priority provider (e.g. CUDA on Linux without `libcuda.so`) is unavailable, the session previously crashed hard. `constructSessions` in `session.js` now tries each execution provider individually, logs a warning for each failure, and continues until one succeeds or all are exhausted. `createInferenceSession` is unchanged.

```js
await pipeline('text-generation', 'onnx-community/LFM2-1.2B-ONNX', { device: 'auto' });
```

### Feature: tokenizer_options passthrough in Text2TextGenerationPipeline (#1096)

```js
const output = await generator('Translate to French: Hello world', {
  max_new_tokens: 100,
  tokenizer_options: { max_length: 64, truncation: true },
});
```

### CODE REVIEW

**Comment by xenova** [comment]
Hi 👋 thanks for the PR. Could you maybe make separate PRs for each issue you are trying to fix? This makes it a lot easier to review.

Thanks

**Comment by Suh0161** [comment]
> Hi 👋 thanks for the PR. Could you maybe make separate PRs for each issue you are trying to fix? This makes it a lot easier to review.
> 
> Thanks

Hi, thanks for the feedback! I've split everything into separate PRs:

- #1671 — Fix num_logits_to_keep default and add get_available_devices() (closes #1666, #1643)
- #1672 — Fix auto device falling back when a provider fails to load (closes #1642)
- #1673 — Allow callers to pass tokenizer_options in Text2TextGenerationPipeline (closes #1096)

The offsets PR is still open separately as well. Closing this one.
